File: website/lib/calculator/math.ts
// ─────────────────────────────────────────────────────────────────────────────
// lib/calculator/math.ts
// Pure math for the /calculator tabs. No dependencies, no side effects — every
// function takes numbers and returns numbers/objects so it can be unit-tested
// and reused by the SVG previews.
// Sikhay and Valiger Collaboration
// ─────────────────────────────────────────────────────────────────────────────

export const TAU = Math.PI * 2

// ── Angle conversion ─────────────────────────────────────────────────────────

export function degToRad(deg: number): number {
  return (deg * Math.PI) / 180
}

export function radToDeg(rad: number): number {
  return (rad * 180) / Math.PI
}

/** Wrap an angle in degrees to [0, 360). */
export function normalizeDeg(deg: number): number {
  return ((deg % 360) + 360) % 360
}

const GREEK_PI = 'π'

/**
 * Express `deg` degrees as an exact multiple of π radians when it lands on a
 * "nice" denominator (2..12), otherwise fall back to a decimal.
 * 90 → "π/2", 180 → "π", 270 → "3π/2", 45 → "π/4", 123 → "0.6833π".
 */
export function toPiFraction(deg: number): string {
  const ratio = deg / 180 // multiples of π
  if (ratio === 0) return '0'

  const sign = ratio < 0 ? '-' : ''
  const abs = Math.abs(ratio)

  for (let denom = 1; denom <= 12; denom++) {
    const num = abs * denom
    if (Math.abs(num - Math.round(num)) < 1e-9) {
      const n = Math.round(num)
      const numPart = n === 1 ? GREEK_PI : `${n}${GREEK_PI}`
      return denom === 1 ? `${sign}${numPart}` : `${sign}${numPart}/${denom}`
    }
  }
  return `${(ratio).toFixed(4)}${GREEK_PI}`
}

// ── Vectors ──────────────────────────────────────────────────────────────────

export interface Vec2 {
  x: number
  y: number
}

export function fromPolar(magnitude: number, angleDeg: number): Vec2 {
  const r = degToRad(angleDeg)
  return { x: magnitude * Math.cos(r), y: magnitude * Math.sin(r) }
}

export function magnitude(v: Vec2): number {
  return Math.hypot(v.x, v.y)
}

/** Angle of a vector in degrees, [0, 360). Returns 0 for the zero vector. */
export function angleDeg(v: Vec2): number {
  if (v.x === 0 && v.y === 0) return 0
  return normalizeDeg(radToDeg(Math.atan2(v.y, v.x)))
}

export interface VectorSum {
  resultant: Vec2
  magnitude: number
  angleDeg: number
}

export function addVectors(a: Vec2, b: Vec2): VectorSum {
  const resultant = { x: a.x + b.x, y: a.y + b.y }
  return {
    resultant,
    magnitude: magnitude(resultant),
    angleDeg: angleDeg(resultant),
  }
}

// ── Rotation ─────────────────────────────────────────────────────────────────

export interface Rotation {
  /** Rotated vector. */
  out: Vec2
  /** The 2×2 rotation matrix, row-major: [[a, b], [c, d]]. */
  matrix: [[number, number], [number, number]]
}

export function rotate(v: Vec2, angleDeg: number): Rotation {
  const t = degToRad(angleDeg)
  const cos = Math.cos(t)
  const sin = Math.sin(t)
  return {
    out: {
      x: v.x * cos - v.y * sin,
      y: v.x * sin + v.y * cos,
    },
    matrix: [
      [cos, -sin],
      [sin, cos],
    ],
  }
}

// ── Polygons ─────────────────────────────────────────────────────────────────

export interface PolygonAngles {
  sides: number
  /** One interior angle of a regular n-gon, degrees. */
  interior: number
  /** One exterior angle, degrees. */
  exterior: number
  /** Sum of all interior angles, degrees. */
  interiorSum: number
  /** Central angle subtended by one side, degrees. */
  central: number
}

export function polygonAngles(sides: number): PolygonAngles {
  const n = Math.max(3, Math.round(sides))
  return {
    sides: n,
    interior: ((n - 2) * 180) / n,
    exterior: 360 / n,
    interiorSum: (n - 2) * 180,
    central: 360 / n,
  }
}

/**
 * Vertices of a regular n-gon inscribed in a circle of radius `r` centered at
 * (cx, cy), first vertex at the top. Uses screen coordinates (+y down).
 */
export function polygonVertices(
  sides: number,
  r: number,
  cx = 0,
  cy = 0,
): Vec2[] {
  const n = Math.max(3, Math.round(sides))
  const verts: Vec2[] = []
  for (let i = 0; i < n; i++) {
    const a = -Math.PI / 2 + (i * TAU) / n
    verts.push({ x: cx + r * Math.cos(a), y: cy + r * Math.sin(a) })
  }
  return verts
}

// ── Formatting ───────────────────────────────────────────────────────────────

/** Trim a float to at most `dp` decimals, dropping trailing zeros. */
export function fmt(n: number, dp = 3): string {
  if (!Number.isFinite(n)) return '—'
  const rounded = Number(n.toFixed(dp))
  return String(rounded)
}

File: app/lib/calculator/math.dart
// ─────────────────────────────────────────────────────────────────────────────
// lib/calculator/math.dart
// Pure math for the Calculator mode's four tabs. Direct port of the website's
// website/lib/calculator/math.ts — every function takes numbers and returns
// numbers/objects, no side effects, so it stays a 1:1 mirror of the TS source.
// Sikhay and Valiger Collaboration
// ─────────────────────────────────────────────────────────────────────────────

import 'dart:math' as math;

const double tau = math.pi * 2;

// ── Angle conversion ─────────────────────────────────────────────────────────

double degToRad(double deg) => (deg * math.pi) / 180;

double radToDeg(double rad) => (rad * 180) / math.pi;

/// Wrap an angle in degrees to [0, 360).
double normalizeDeg(double deg) => ((deg % 360) + 360) % 360;

const String _greekPi = 'π';

/// Express [deg] degrees as an exact multiple of π radians when it lands on a
/// "nice" denominator (2..12), otherwise fall back to a decimal.
/// 90 → "π/2", 180 → "π", 270 → "3π/2", 45 → "π/4", 123 → "0.6833π".
String toPiFraction(double deg) {
  final ratio = deg / 180; // multiples of π
  if (ratio == 0) return '0';

  final sign = ratio < 0 ? '-' : '';
  final abs = ratio.abs();

  for (int denom = 1; denom <= 12; denom++) {
    final num = abs * denom;
    if ((num - num.round()).abs() < 1e-9) {
      final n = num.round();
      final numPart = n == 1 ? _greekPi : '$n$_greekPi';
      return denom == 1 ? '$sign$numPart' : '$sign$numPart/$denom';
    }
  }
  return '${ratio.toStringAsFixed(4)}$_greekPi';
}

// ── Vectors ──────────────────────────────────────────────────────────────────

class Vec2 {
  final double x;
  final double y;
  const Vec2(this.x, this.y);
}

Vec2 fromPolar(double magnitude, double angleDeg) {
  final r = degToRad(angleDeg);
  return Vec2(magnitude * math.cos(r), magnitude * math.sin(r));
}

double vecMagnitude(Vec2 v) => math.sqrt(v.x * v.x + v.y * v.y);

/// Angle of a vector in degrees, [0, 360). Returns 0 for the zero vector.
double vecAngleDeg(Vec2 v) {
  if (v.x == 0 && v.y == 0) return 0;
  return normalizeDeg(radToDeg(math.atan2(v.y, v.x)));
}

class VectorSum {
  final Vec2 resultant;
  final double magnitude;
  final double angleDeg;
  const VectorSum({required this.resultant, required this.magnitude, required this.angleDeg});
}

VectorSum addVectors(Vec2 a, Vec2 b) {
  final resultant = Vec2(a.x + b.x, a.y + b.y);
  return VectorSum(
    resultant: resultant,
    magnitude: vecMagnitude(resultant),
    angleDeg: vecAngleDeg(resultant),
  );
}

// ── Rotation ─────────────────────────────────────────────────────────────────

/// The 2×2 rotation matrix, row-major: [[a, b], [c, d]].
class Mat2 {
  final double a, b, c, d;
  const Mat2(this.a, this.b, this.c, this.d);
}

class RotationResult {
  final Vec2 out;
  final Mat2 matrix;
  const RotationResult({required this.out, required this.matrix});
}

RotationResult rotateVec(Vec2 v, double angleDeg) {
  final t = degToRad(angleDeg);
  final cos = math.cos(t);
  final sin = math.sin(t);
  return RotationResult(
    out: Vec2(v.x * cos - v.y * sin, v.x * sin + v.y * cos),
    matrix: Mat2(cos, -sin, sin, cos),
  );
}

// ── Polygons ─────────────────────────────────────────────────────────────────

class PolygonAngles {
  final int sides;
  /// One interior angle of a regular n-gon, degrees.
  final double interior;
  /// One exterior angle, degrees.
  final double exterior;
  /// Sum of all interior angles, degrees.
  final double interiorSum;
  /// Central angle subtended by one side, degrees.
  final double central;
  const PolygonAngles({
    required this.sides,
    required this.interior,
    required this.exterior,
    required this.interiorSum,
    required this.central,
  });
}

PolygonAngles polygonAngles(double sides) {
  final n = math.max(3, sides.round());
  return PolygonAngles(
    sides: n,
    interior: ((n - 2) * 180) / n,
    exterior: 360 / n,
    interiorSum: ((n - 2) * 180).toDouble(),
    central: 360 / n,
  );
}

/// Vertices of a regular n-gon inscribed in a circle of radius [r] centered at
/// (cx, cy), first vertex at the top. Uses screen coordinates (+y down).
List<Vec2> polygonVertices(double sides, double r, {double cx = 0, double cy = 0}) {
  final n = math.max(3, sides.round());
  final verts = <Vec2>[];
  for (int i = 0; i < n; i++) {
    final a = -math.pi / 2 + (i * tau) / n;
    verts.add(Vec2(cx + r * math.cos(a), cy + r * math.sin(a)));
  }
  return verts;
}

// ── Formatting ───────────────────────────────────────────────────────────────

/// Trim a float to at most [dp] decimals, dropping trailing zeros.
String fmt(double n, [int dp = 3]) {
  if (!n.isFinite) return '—';
  final factor = math.pow(10, dp);
  final rounded = (n * factor).round() / factor;
  // Match JS `String(Number(n.toFixed(dp)))` — trims trailing zeros/decimal point.
  var s = rounded.toStringAsFixed(dp);
  if (s.contains('.')) {
    s = s.replaceFirst(RegExp(r'0+$'), '');
    s = s.replaceFirst(RegExp(r'\.$'), '');
  }
  if (s == '-0') s = '0';
  return s;
}
